### ISSUE TITLE
CFFI Version Dependency

### ISSUE THREAD

**Issue by SpencerRP** [innobi/pantab#21]
Running this package from fresh installs of Anaconda 3.6 and the tableauhyperapi results in the error message below. Doing a pip install of msgpack and then of cffi resolves this issue. It is related to the pycparser version, but I can't tell you which version is needed as I didn't think to save the relevant logs.

```
 File "<ipython-input-3-62d421bcaed4>", line 1, in <module>
   extract_queries('Y:/UiPath/Tableau Extract/','Hyper_Refresh.xlsx')

 File "<ipython-input-1-44a04ab0d533>", line 92, in extract_queries
   process_query(path, filename, connection, sql)

 File "<ipython-input-1-44a04ab0d533>", line 74, in process_query
   pantab.frame_to_hyper(combined_csv, top + '\\' + filename, table = filename.rsplit('.')[0])

 File "C:\ProgramData\Anaconda3\lib\site-packages\pantab\_pantab.py", line 157, in frame_to_hyper
   with HyperProcess(Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU) as hpe:

 File "C:\ProgramData\Anaconda3\lib\site-packages\tableauhyperapi-0.0.8707-py3.6.egg\tableauhyperapi\hyperprocess.py", line 83, in __init__
   InteropUtil.string_to_char_p(user_agent),

 File "C:\ProgramData\Anaconda3\lib\site-packages\tableauhyperapi-0.0.8707-py3.6.egg\tableauhyperapi\impl\dllutil.py", line 34, in string_to_char_p
   return ffi.from_buffer('char[]', s.encode())

TypeError: from_buffer() takes 2 positional arguments but 3 were given```

**Comment by WillAyd** [comment]
Thanks for the report. Can you provide sample data to reproduce?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by WillAyd** [comment]
Also note that there is a newer version of the Hyper API available. Would be curious to know if you see the same error when upgrading that

**Comment by WillAyd** [comment]
Closing as I don’t think this is an issue with pantab, but if unresolved on your end still lmk

**Comment by SpencerRP** [comment]
Sorry for being slow to reply, I didn't check my account for a while.

You're correct. Reviewing the error message in my original post, it's an
issue with tableauhyperapi depending on a particular version of pycparser
(via msgpack, via cffi), rather than pantab. I went to you because I ran
into it using your package and am unfamiliar with the api's build method.
(I'm kind of a newb, and lack a formal education.) Do you know if the api's
build method can handle specific dependency versions?

P.S. Pantab is handy, thanks for making it.

**Comment by WillAyd** [comment]
Glad you find pantab helpful! With regards to the tableauhyperapi, I can't say definitely what msgpack / cffi requirements it may have. If you can isolate problematic version(s) that would be helpful, or alternately could try asking on Tableau forums

**Comment by SpencerRP** [comment]
Tomorrow I'll double-check the version of anaconda I installed, but I
believe it's 5.2.0 as I went off this post to find the latest distribution
that included Python 3.6 (
https://stackoverflow.com/questions/54801513/how-can-i-download-anaconda-for-python-3-6).
The problematic versions in this are cffi 1.11.5, msgpack-python 0.5.6,
pycparser 2.18. Updating these to the latest version in order from last to
first resolved the issue. I was hoping to make it so that people who use a
similar setup don't need to spend ~2 hours googling around to fix the issue
on their end.

**Comment by roediger** [comment]
I am checking if this is an issue with tableauhyperapi that I could fix. However, I failed to repro this with a fresh Anaconda-5.2.0 installation. Do you have more specific repro steps?

**Comment by SpencerRP** [comment]
I’ll try reproducing it some time today and get back to you.

**Comment by SpencerRP** [comment]
Wasn't able to replicate the issue with a fresh Anaconda3-5.2.0-Windows-x86_64 install using a pip installed tableauhyperapi-0.0.8953-py3-none-win_amd64.whl, an a "pip install pantab" command that led to pantab-0.1.0 being installed, as well as data that would be representative of what I used when I first encountered the issue.

There are two differences from what I recall:
A) I believe that I installed it using a zip file rather than a whl file.
B) I don't believe I saw a message for the pantab install saying "distributed 1.21.8 requires msgpack, which is not installed."

Not sure if these would make a difference. I'll mess around with it more at some point, but I don't consider it a priority anymore as a) it works with the whl file install b) the whl file install is what your install instructions strongly emphasize.
